src/tiledbsoma_ml/io_batches.py

```diff
@@ -0,0 +1,139 @@
+# Copyright (c) 2021-2024 The Chan Zuckerberg Initiative Foundation
+# Copyright (c) 2021-2024 TileDB, Inc.
+#
+# Licensed under the MIT License.
+
+import gc
+import logging
+import time
+from typing import Iterable, Iterator, Optional, Sequence, Tuple
+
+import attrs
+import numpy as np
+import pandas as pd
+import pyarrow as pa
+from tiledbsoma import DataFrame, IntIndexer, SparseNDArray
+
+from tiledbsoma_ml._csr import CSR_IO_Buffer
+from tiledbsoma_ml._utils import EagerIterator, batched
+from tiledbsoma_ml.common import NDArrayJoinId
+from tiledbsoma_ml.query_ids import Chunks
+
+logger = logging.getLogger("tiledbsoma_ml.io_batches")
+IOBatch = Tuple[CSR_IO_Buffer, pd.DataFrame]
+"""Tuple type emitted by |IOBatches|, containing ``X`` rows (as a |CSR_IO_Buffer|) and ``obs`` rows (|pd.DataFrame|)."""
+
+
+@attrs.define(frozen=True)
+class IOBatches(Iterable[IOBatch]):
+    """Given a list of ``obs_joinid`` |Chunks|, re-chunk them into (optionally shuffled) |IOBatch|'s".
+
+    An |IOBatch| is a tuple consisting of a batch of rows from the ``X`` |SparseNDArray|, as well as the corresponding
+    rows from the ``obs`` |DataFrame|. The ``X`` rows are returned in an optimized |CSR_IO_Buffer|.
+    """
+
+    chunks: Chunks
+    io_batch_size: int
+    obs: DataFrame
+    var_joinids: NDArrayJoinId
+    X: SparseNDArray
+    obs_column_names: Sequence[str] = ("soma_joinid",)
+    seed: Optional[int] = None
+    shuffle: bool = True
+    use_eager_fetch: bool = True
+
+    @property
+    def io_batch_ids(self) -> Iterable[Tuple[int, ...]]:
+        """Re-chunk ``obs_joinids`` according to the desired ``io_batch_size``."""
+        return batched(
+            (joinid for chunk in self.chunks for joinid in chunk),
+            self.io_batch_size,
+        )
+
+    def __iter__(self) -> Iterator[IOBatch]:
+        """Emit |IOBatch|'s."""
+        # Create RNG - does not need to be identical across processes, but use the seed anyway
```

Comment:
I had to think about this comment a bit.  It would be more helpful to our future selves if it said something like:   Because obs/var IDs are pre-partitioned and shuffled, this RNG does not need to be identical across sub-processes.